rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,630 @@
+   considered to be Undefined Behavior, and may therefore result in unintended
+   compilation output if detected by the optimizer.
+2. Using an overly wide volatile load or store operation which cannot be carried
+   out by a single hardware load and store instruction will not result in a
+   compilation error, but in the silent emission of multiple hardware load or
+   store instructions, which might be a logic error in the users' program.
+
+By implementing support for LLVM's atomic volatile operations, and encouraging
+their use on every hardware that supports them, we eliminate these divergences
+from hardware behavior and therefore bring volatile operations closer to the
+"defer to hardware memory model" semantics that programmers expect them to have.
+This reduces the odd of mistake in programs operating outside of the regular
+Rust Abstract Machine semantics, which are notoriously hard to get right.
+
+As an unexpected but attractive side-effect, atomic volatile memory operations
+also enable higher-performance interprocess communication between mutually
+trusting Rust programs, through lock-free synchronization of shared memory.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler generally assumes that the program that it is building is
+living in a fully isolated memory space. It leverages this knowledge to
+transform said program's memory access patterns for performance optimization
+purposes, under the assumption that said transformations will not have any
+externally observable effect other than the program running faster.
+
+Examples of such transformations include:
+
+- Caching data from RAM into CPU registers.
+- Eliminating unused loads and unobservable stores.
+- Merging neighboring stores with each other.
+- Only updating the part of an over-written struct that has actually changed.
+
+Although these optimizations are most of the time correct and useful, there are
+some situations where they are inappropriate, including but not limited to:
+
+- [Memory-mapped I/O](https://en.wikipedia.org/wiki/Memory-mapped_I/O), a common
+  low-level communication protocol between CPUs and peripherals, where hardware
+  registers masquerading as memory can be used to program peripherals by
+  accessing said registers in very specific load/store patterns.
+- [Shared memory](https://en.wikipedia.org/wiki/Shared_memory), a form of
+  inter-process communication where two programs can communicate via a common
+  memory block, and it is therefore not appropriate for Rust to assume that it
+  is aware of all memory accesses occurring inside said memory block.
+- [Cryptography](https://en.wikipedia.org/wiki/Cryptography), where it is
+  extremely important to ensure that sensitive information is erased after use,
+  and is not leaked via indirect means such as recognizable scaling patterns in
+  the time taken by a system to process attacker-crafted inputs.
+
```

Comment:
I would add an additional use case here: speculative loads of data that may be concurrently modified. This is the basic idea behind [seqlocks](https://github.com/Amanieu/seqlock): first attempt a read, then check if the data was modified while we were reading it. If there was a concurrent modification, discard the read data and retry.

Reply:
Seqlocks don't need volatile, and trying to use volatile to fix seqlocks in C is IMO a bad hack.

Seqlocks need a way to do a non-atomic read such that data races are not UB, but return uninit/undef. Basically, they need the LLVM memory model.